### ISSUE TITLE
Emulate Keypresses?

### ISSUE THREAD

**Issue by schmolmo** [munshkr/flok#201]
Hey there,

I've got a question that might be out of scope for a github issue - if so, please feel free to close/delete it!
Would it be possible to emulate keypresses (insert, delete, move cursors) via the Javascript console from a single client?
My use case would be the following: I have a system with multiple keyboards (for multiple users) connected and would like to use emulate a flok session with multiple users, but only on one computer.

Could you maybe point me in a direction? Help would be greatly appreciated!

All best,
moritz

**Comment by munshkr** [comment]
Hey! It's not the first time I was asked to have a way to programmatically control the editor, maybe instead of emulating keypresses, you could just add characters and move the cursor of each editor via JavaScript.

The problem I see is that if these keyboards are connected directly to the same computer, they will all send the same keystrokes and there would be no way to distinguish each other. Or are they connected to some kind of hardware or something? Is this for some kind of installation?

**Comment by schmolmo** [comment]
Thank you for the response! Sorry, I should have been more detailed: Yes, the (old PS/2) keyboards are connected to Arduino-based controllers (one controller per keyboard) that are then connected to the Computer via MIDI. I can distinguish the keyboards in the little MIDI-based protocol I wrote. 
I could supply commands like  ```Insert Character A for User 02``` or ```Move Cursor left for User 04```. I probably would have a program in between to translate these messages. Do you think that there could be an easy solution to make flok react to those messages?
The aim is to have a one-computer-multi-user-livecoding scenario, more or less :)

**Comment by munshkr** [comment]
Ahhh interesting! That makes sense :)

Yes, I think it should be possible. I can think of an API to control Flok from a parent window via [Window messaging](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage). In the pre-1.0 version I had something like this for evaluating messages. This way, you could embed Flok in an IFrame, and control everything through WebMIDI/WebSockets, and your multikeyboard controller...

**Comment by schmolmo** [comment]
That sounds very interesting, thank you! As far as I understand, I wouldn't have to make any changes to flok, but only write a system with the IFrames and the Message relaying, right? Is there any way to join a session under a specific nickname?

**Comment by munshkr** [comment]
Exactly, you can just iframe https://next.flok.cc with some query parameters, e.g. https://next.flok.cc/s/my-session-1234?username=munshkr&targets=hydra,hydra,hydra,hydra

**Comment by schmolmo** [comment]
Great, thank you! When I use the link you provided, the dialog for configuring targets + username still comes up. Is there anyway to get around this? 
And (sorry for all the questions): Is there a resource (or some lines in the flok source) that could give me a clue which calls i could send from the iframe parent?

**Comment by munshkr** [comment]
I haven't deployed those query parameters yet (still working on that! https://github.com/munshkr/flok/pull/200), but don't worry, I'll soon merge that and document properly. I'll update you here.

**Comment by schmolmo** [comment]
Thank you very much, this sounds great! :-)
